platform/linux-generic/odp_packet_io.c

```diff
@@ -2472,3 +2485,345 @@ int odp_pktout_ts_read(odp_pktio_t hdl, odp_time_t *ts)
+
+	payload_len = lso_opt->max_payload_len;
+	hdr_len     = lso_opt->payload_offset;
+	pkt_len     = odp_packet_len(packet);
+	pkt_payload = pkt_len - hdr_len;
+
+	if (hdr_len > pkt_len || payload_len > pkt_payload) {
+		ODP_ERR("Too short packet\n");
+		return -1;
+	}
+
+	pool = odp_packet_pool(packet);
+	pkt_len = hdr_len + payload_len;
+	num_pkt = pkt_payload / payload_len;
+	num_full = num_pkt;
+	left_over_len = pkt_payload - (num_pkt * payload_len);
+	if (left_over_len) {
+		left_over = 1;
+		num_pkt++;
+	}
+
+	/* Alloc packets */
+	odp_packet_t pkt_out[num_pkt];
+
+	num = odp_packet_alloc_multi(pool, pkt_len, pkt_out, num_full);
+	if (num < num_full) {
+		ODP_DBG("Alloc failed %i\n", num);
+		if (num > 0) {
+			num_free = num;
+			goto error;
+		}
+	}
+
+	if (left_over) {
+		pkt = odp_packet_alloc(pool, hdr_len + left_over_len);
+		if (pkt == ODP_PACKET_INVALID) {
+			ODP_DBG("Alloc failed\n");
+			num_free = num_full;
+			goto error;
+		}
+
+		pkt_out[num_pkt - 1] = pkt;
+	}
+
+	/* Copy headers */
+	for (i = 0; i < num_pkt; i++) {
+		if (odp_packet_copy_from_pkt(pkt_out[i], 0, packet, 0, hdr_len)) {
+			ODP_ERR("Header copy failed\n");
+			num_free = num_pkt;
+			goto error;
+		}
+	}
+
+	/* Copy payload */
+	for (i = 0; i < num_full; i++) {
+		offset = hdr_len + (i * payload_len);
+		if (odp_packet_copy_from_pkt(pkt_out[i], hdr_len, packet, offset, payload_len)) {
+			ODP_ERR("Payload copy failed\n");
+			num_free = num_pkt;
+			goto error;
+		}
+	}
+
+	if (left_over) {
+		offset = hdr_len + (num_full * payload_len);
+		if (odp_packet_copy_from_pkt(pkt_out[num_pkt - 1], hdr_len, packet, offset,
+					     left_over_len)){
+			ODP_ERR("Payload copy failed\n");
+			num_free = num_pkt;
+			goto error;
+		}
+	}
+
+	/* Update custom fields */
+	for (i = 0; num_custom && i < num_pkt; i++) {
+		if (lso_update_custom(lso_prof, pkt_out[i], i)) {
+			ODP_ERR("Custom field update failed. Segment %i\n", i);
+			num_free = num_pkt;
+			goto error;
+		}
+	}
+
+	ret = odp_pktout_send(queue, pkt_out, num_pkt);
+
+	if (ret < num_pkt) {
+		ODP_DBG("Packet send failed %i\n", ret);
+		num_free = num_pkt;
+		if (ret > 0) {
+			first_free = ret;
+			num_free = num_pkt - ret;
+		}
+	}
+
+	odp_packet_free(packet);
+
+	return 0;
+
+error:
+	odp_packet_free_multi(&pkt_out[first_free], num_free);
+	return -1;
+}
+
+int odp_pktout_send_lso(odp_pktout_queue_t queue, const odp_packet_t packet[], int num,
+			const odp_packet_lso_opt_t *opt)
+{
+	int i;
+	odp_packet_t pkt;
+	odp_packet_lso_opt_t lso_opt;
+	const odp_packet_lso_opt_t *opt_ptr = &lso_opt;
+
+	if (odp_unlikely(num <= 0)) {
+		ODP_ERR("No packets\n");
+		return -1;
+	}
+
+	memset(&lso_opt, 0, sizeof(odp_packet_lso_opt_t));
+	if (opt)
+		opt_ptr = opt;
+
+	for (i = 0; i < num; i++) {
+		pkt = packet[i];
+
+		if (opt == NULL) {
+			odp_packet_hdr_t *pkt_hdr = packet_hdr(pkt);
+
+			if (pkt_hdr->p.flags.lso == 0) {
+				ODP_DBG("No LSO options on packet %i\n", i);
```

Comment:
This is user error, so ODP_ERR() could be used.

Reply:
fixed